File: app/src/main/java/com/bitflaker/lucidsourcekit/charts/CircleGraph.java
package com.bitflaker.lucidsourcekit.charts;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.view.View;

import com.bitflaker.lucidsourcekit.R;
import com.bitflaker.lucidsourcekit.general.Tools;

public class CircleGraph extends View {
    private int val1;
    private int val2;
    private float lineWidth;
    private final Paint dataLinePaint = new Paint();
    private final Paint dataLinePaint2 = new Paint(Paint.ANTI_ALIAS_FLAG | Paint.DITHER_FLAG);
    private final Paint dataLinePaintBackgroundEraser = new Paint();
    private final Paint dataLinePaintSpaceEraser = new Paint();
    private final RectF rf = new RectF();

    public CircleGraph(Context context, AttributeSet as){
        super(context, as);
        setLayerType(View.LAYER_TYPE_SOFTWARE, null);
        dataLinePaint.setColor(Tools.getAttrColor(R.attr.colorSecondary, getContext().getTheme()));
        dataLinePaint.setAntiAlias(true);
        dataLinePaint2.setColor(Tools.getAttrColor(R.attr.colorPrimary, getContext().getTheme()));
        dataLinePaint2.setAntiAlias(true);
        dataLinePaintBackgroundEraser.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.CLEAR));
        dataLinePaintBackgroundEraser.setAntiAlias(true);
        dataLinePaintSpaceEraser.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.CLEAR));
        dataLinePaintSpaceEraser.setAntiAlias(true);
        val1 = 2;
        val2 = 3;
        lineWidth = Tools.dpToPx(context, 15);
        dataLinePaintSpaceEraser.setStrokeWidth(Tools.dpToPx(getContext(), Tools.dpToPx(context, 1.25)));
    }

    public void setData(int val1, int val2, float lineWidth, float space) {
        this.val1 = val1;
        this.val2 = val2;
        this.lineWidth = lineWidth;
        dataLinePaintSpaceEraser.setStrokeWidth(Tools.dpToPx(getContext(), space));
        invalidate();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        final float height = getHeight();
        final float width = getWidth();
        final float radius = width > height ? height * 0.5f : width * 0.5f;
        final float centerX = getWidth() * 0.5f;
        final float centerY = getHeight() * 0.5f;
        final float degVal1 = (360.0f * val1) / (val1 + val2);

        rf.set(centerX - radius, centerY - radius, centerX + radius, centerY + radius);

        canvas.drawArc(rf, 270, degVal1, true, dataLinePaint);
        canvas.drawArc(rf, 270 + degVal1, 360 - degVal1, true, dataLinePaint2);

        if(val1 != 0 && val2 != 0){
            double angle = 270 + degVal1;       // degrees
            angle = angle * Math.PI / 180.0;    // radians
            double xVal = centerX + Math.cos(angle) * centerX;
            double yVal = centerY + Math.sin(angle) * centerY;
            canvas.drawLine(centerX, centerY, centerX, 0, dataLinePaintSpaceEraser);
            canvas.drawLine(centerX, centerY, (float)xVal, (float)yVal, dataLinePaintSpaceEraser);
        }

        canvas.drawCircle(centerX, centerY, radius - lineWidth, dataLinePaintBackgroundEraser);
    }
}

File: app/src/main/java/com/bitflaker/lucidsourcekit/views/CircleGraph.kt
package com.bitflaker.lucidsourcekit.views

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.RectF
import android.util.AttributeSet
import android.view.View
import com.bitflaker.lucidsourcekit.R
import com.bitflaker.lucidsourcekit.utils.attrColor
import com.bitflaker.lucidsourcekit.utils.dpToPx
import kotlin.math.cos
import kotlin.math.sin

class CircleGraph(context: Context, attributeSet: AttributeSet?) : View(context, attributeSet) {
    var value1: Int = 2
    var value2: Int = 3
    var lineWidth: Float = 15.0f.dpToPx
    var separatorWidth: Float = 1.25f.dpToPx
    private val startAngle: Float = 270f
    private val bounds = RectF()
    private val dataLinePaint = Paint(Paint.ANTI_ALIAS_FLAG or Paint.DITHER_FLAG)
    private val dataLinePaintEraser = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        xfermode = PorterDuffXfermode(PorterDuff.Mode.CLEAR)
    }

    init {
        setLayerType(LAYER_TYPE_SOFTWARE, null)
    }

    fun setData(value1: Int, value2: Int, lineWidth: Float, separatorWidth: Float) {
        this.value1 = value1
        this.value2 = value2
        this.lineWidth = lineWidth
        this.separatorWidth = separatorWidth
        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        // Cache current view dimensions
        val height = height
        val width = width

        // Get the radius of the shortest side and the center to fit the circle in the view
        val radius = if (width > height) height * 0.5f else width * 0.5f
        val centerX = width * 0.5f
        val centerY = height * 0.5f

        // Get the sweep angle of both values
        val sweepValue1 = (360.0f * value1) / (value1 + value2)
        val sweepValue2 = 360 - sweepValue1

        // Set the bounds used for the arcs
        bounds.set(centerX - radius, centerY - radius, centerX + radius, centerY + radius)

        // Draw both value arcs
        dataLinePaint.color = context.attrColor(R.attr.colorTertiary)
        canvas.drawArc(bounds, startAngle, sweepValue1, true, dataLinePaint)
        dataLinePaint.color = context.attrColor(R.attr.colorPrimary)
        canvas.drawArc(bounds, startAngle + sweepValue1, sweepValue2, true, dataLinePaint)

        // Draw a separator line between both values
        if (value1 != 0 && value2 != 0) {
            val angle = (startAngle + sweepValue1) * Math.PI / 180.0
            val xVal = centerX + cos(angle).toFloat() * centerX
            val yVal = centerY + sin(angle).toFloat() * centerY

            // Draw separator lines with eraser
            dataLinePaintEraser.strokeWidth = separatorWidth
            canvas.drawLine(centerX, centerY, centerX, 0f, dataLinePaintEraser)
            canvas.drawLine(centerX, centerY, xVal, yVal, dataLinePaintEraser)
        }

        // Clear the center of the circle to be left with only the outlines
        dataLinePaintEraser.strokeWidth = 0f
        canvas.drawCircle(centerX, centerY, radius - lineWidth, dataLinePaintEraser)
    }
}